### DISCUSSION TITLE
How to deny users access to table while allowing access to view?

### DISCUSSION THREAD

**Discussion by gianmarcocalbi** [supabase/supabase#28533] [Questions] [answered]
I have a table `clients` with `id`, `name`, `email` and I want authenticated users not to be able to see investors' `id`.

I have created a **view** that contains only `name` and `email`.

There is a RLS policy that allows a user to select only their clients. If I understood correctly by setting `security_invoker = true` when creating the view, the view will inherit the RLS of the tables it selects data from, thus a user will be able to select rows only of their clients from the view as well.

Now I would like to deny users access to the `clients` table while allowing them to select from the view only as suggested in the [documentation](https://supabase.com/docs/guides/database/tables?queryGroups=database-method&database-method=sql#security):

> **Security**
Views can restrict the amount and type of data presented to a user. Instead of allowing a user direct access to a set of tables, we provide them a view instead. We can prevent them from reading sensitive columns by excluding them from the underlying query.

My question is about the correct way to implement this: "Instead of allowing a user direct access to a set of tables, we provide them a view instead".

Should I put `clients` in a private schema while the view in the public one?

Thank you

**Comment by GaryAustin1** [comment, marked as answer]
Seems like you could do that.  Not sure how that complicates the rest of your code keeping that updated.

You could also do a security definer view and put all your RLS in the where statement of the view.  You can use auth.uid() there but NOT RLS on any tables.

You can also use Postgres column privileges to not allow authenticated to select that column.

**Reply by gianmarcocalbi** [reply]
Okay. I will go for the second option you propose: I will keep the tables in the `public` schema but define no RLS policies so they won't be accessible, while defining views with RLS in the where statement. 

Thank you

**Reply by gianmarcocalbi** [reply]
The only "minor" issue I am getting with this approach is that views will have the SECURITY DEFINER property, because the RLS is defined in the where clause and it cannot be inherited from the table they select from, otherwise users won't be able to query them. So I getting 7 **severe errors** in the security advisor

**Reply by GaryAustin1** [reply]
I would generate an issue in supabase/supabase github on how to silence the error if you know you have it covered.  Right now there are many cases (like for a searchpath not being '' in a security definer function) that even though you may have things covered you get an error.
